### ISSUE TITLE
SyslogTcpSink use of SslProtocols.None is invalid

### ISSUE THREAD

**Issue by wldaniel1** [IonxSolutions/serilog-sinks-syslog#47]
According to .NET documentation, SslProtocols.None is recommended to allow the OS to chose the best protocol.  SyslogTcpSink is using it as a flag to enable/disable TLS.  There should be another flag in SyslogTcpConfig to control TLS vs. plain text instead of using SslProtocols.None.

**Comment by colin-anderson** [comment]
Ah, yes, I guess you mean [this](https://docs.microsoft.com/en-gb/dotnet/framework/network-programming/tls)?

Hmm, that's annoying 😕   With hindsight, we should have used a separate boolean toggle `SyslogTcpConfig.TlsEnabled`, or perhaps even made `SyslogTcpConfig.SecureProtocols` nullable (with `null` indicating disabled TLS).

@wldaniel1 @Kevin-Andrew any ideas if we can work around this without a breaking change?

Only idea I have is to add a new `bool? TlsEnabled` flag - null or false would both indicate disabled, but because it's nullable we could default it to `true` if anything other than `SslProtocols.None` was used. I think that should keep things working without change for existing users, but having a single nullable property feels a bit icky 🤢

**Comment by wldaniel1** [comment]
@colin-anderson yes, that's what I meant.

I think your solution makes sense.  It could be a nullable parameter instead of nullable property.  

![image](https://user-images.githubusercontent.com/98424697/151200438-cad9d575-887c-4d96-b7fe-b6db4dd54b42.png)

**Comment by Kevin-Andrew** [comment]
@colin-anderson @wldaniel1 

I think you'd just be introducing more confusion. The documentation on the component mentions what `None` does, in the context of this component.
![image](https://user-images.githubusercontent.com/22459191/151215280-cffc1f45-8b4e-4896-ac2a-4d63e8f410b1.png)
and
![image](https://user-images.githubusercontent.com/22459191/151215372-7f4ec0f9-b507-4157-8125-e6189fe3514b.png)

I don't disagree with you in that by exposing the `SslProtocols` type, it should be used in a manor consistent with it's documentation, but also note that the Intellisense documentation on `SslProtocols.None` does not match the MSDN online documentation and could therefore be interpreted different ways:
![image](https://user-images.githubusercontent.com/22459191/151215968-29978259-5a08-4f4a-bf40-d0011d61f74b.png)

So to me, that is 3 places where the intention/purpose of `None` is pretty clear. While @wldaniel1 knows the real intent/purpose of `None`, as it is used within the .NET Framework, I'm not sure most other people do. I didn't even bother to think about it or look it up when I first used this component.

So if we try to preserve backwards compatibility by introducing another config setting which then gives dual meaning to `None`, at what point would you break backwards compatibility? The next major release, like 3.0? If so, then why not just wait and change the behavior then, doing it in such a way that forces a compile time error, hopefully making the new usage clear.

While the proposed change would preserve backwards compatibility, like I said, I would be worried about the confusion it may cause. For example, what would the behavior be if the new `UseTLS` property was explicitly set to `false`, but the `SecureProtocols` was set to `SslProtocols.Tls12`? The way I see it is that not only would the documentation for `None` have to change and be given a dual meaning, we would have to also document the fact that if you explicitly set `UseTLS` to `false`, it completely ignores the value of `SecureProtocols`, **which someone else may consider a change in behavior**.

Let me throw this out here as a possible alternative. What if the component declared and used its own `SslProtocols` enum? And for `None`, it explicitly stated that no SSL would be used at all. Would you argue that it is still wrong? Just because we happened to have used the same name?

It appears that making that change may not require any consumers to change any of their code (unless they used a fully qualified `SslProtocols` somewhere):
![image](https://user-images.githubusercontent.com/22459191/151272829-0c0698cb-f038-4c15-8fe6-db107074a476.png)

If the concern is that you can't get the:
> Allows the operating system to choose the best protocol to use,..

aspect of the .NET Framework's `SslProtocols.None`, then I believe the workaround for that is to simply specify all enum values OR'ed together:
```
SslProtocols.Ssl2 | SslProtocols.Ssl3 | SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12 | SslProtocols.Tls13
```
Then the component will use SSL and the server will pick the _best_ protocol, which would give you the same behavior as if we were to introduce a new property for `UseTLS` and you specify `SslProtocols.None`.

I'd be happy to try and make the change and write the documentation if you feel really strong about it. But given that, technically speaking, there is nothing wrong/not possible with the current implementation, that is, you can still get the same functionality of the .NET Framework's definition of `SslProtocols.None`, I think that if any time were to be spent on it, it should be for the next version if you wanted to make it into a breaking change.

**Comment by colin-anderson** [comment]
@Kevin-Andrew yep, IMO Microsoft should have introduced an `SslProtocols.Auto`, which would have made the intent clear.

> What if the component declared and used its own SslProtocols enum

Damn, that's a good idea! 👍  We could do something like (note the possible addition of an "Auto" value):

```c#
[Flags]
public enum SslProtocols
{
	Auto = -1, // <-- we'd map this to System.Security.Authentication.SslProtocols.None
	None = 0, // <-- Don't use SSL/TLS
	Ssl2 = 12, // 0x0000000C
	Ssl3 = 48, // 0x00000030
	Tls = 192, // 0x000000C0
	Tls11 = 768, // 0x00000300
	Tls12 = 3072, // 0x00000C00
	Tls13 = 12288 // 0x00003000
}
```

> then I believe the workaround for that is to simply specify all enum values OR'ed together:

Hmm, not sure this is right - `None` means "let the OS choose", not "use any protocol, even ones explicitly disabled in the OS". Perhaps adding a new `Auto` value to your proposed `SslProtocols` enum would be a better solution?

**Comment by wldaniel1** [comment]
@Kevin-Andrew @colin-anderson 

I agree that .NET's naming is poor.  They should have used Auto like you said or they should have called Default All and made Default behave like None.  

However, I disagree with the argument about the useTLS flag adding confusion.  Your first point was about the project's documentation, which can easily be changed.  

And my version of VS shows this for None.

![image](https://user-images.githubusercontent.com/98424697/151365509-c2516e9b-9a97-45d6-aced-987c9a47bc8e.png)

So after updating documentation, there wouldn't be any dual definition.

> While the proposed change would preserve backwards compatibility, like I said, I would be worried about the confusion it may cause. For example, what would the behavior be if the new UseTLS property was explicitly set to false, but the SecureProtocols was set to SslProtocols.Tls12?

If you aren't doing TLS, then SecureProtocols should be irrelevant.  Same as if you set SecureProtocols to Tls12 and provided a client certificate.  

> Let me throw this out here as a possible alternative. What if the component declared and used its own SslProtocols enum? And for None, it explicitly stated that no SSL would be used at all. Would you argue that it is still wrong? Just because we happened to have used the same name?

I wouldn't say it's wrong, just a little weird.  If this was being written from scratch with the current .NET use of None, would this still be the design?  

Let me just say that I recently integrated this package.  I followed the example here

![image](https://user-images.githubusercontent.com/98424697/151366796-d2b583c7-1fa6-4fb8-9735-d002087d5e85.png)

The part about None is chopped off.  VS didn't display any intellisense for SecureProtocols, so I changed SecureProtocols to None as I have been doing in all my code.  I want my application to switch to TLS 1.3/1.4/2.0 as they become available without deploying code changes.  It took me some time to figure out why things weren't working.  I wasn't sure if it was the client side or the test server I stood up.  After some back and forth I went to the code here to find the issue I was having.   

I am currently setting SecureProtocols to SslProtocols.Tls12 | SslProtocols.Tls13 as a workaround.  I believe introducing another SslProtocols enum would cause my code to not compile.  Which is ok, I can fix it, but this change would not be backwards compatible. 

And @colin-anderson is right, you'd have to map it to SslProtocols.None instead of ORing everything together (that is basically SslProtocols.Default).  With None, the OS (on Windows) looks at the SCHANNEL settings in the registry.

In summary: I don't need anything changed urgently.  I have a suitable workaround specifying the 2 SslProtocols I want to allow.  Anything lower than 1.2 is no longer secure and shouldn't be used.  I would like some option in the future to get None passed to the actual .NET SslStream.  That way OS updates take care of what is allowed.  I'm not picky about how that is accomplished, I could make either solution work for me.

**Comment by Kevin-Andrew** [comment]
@wldaniel1 , for the Intellisense documentation of `SslProtocols.None`, I looked more and found:

.NET Core/5/6: New documentation
.NET Framework: Old documentation
.NET Standard 2.0: Old documentation
.NET Standard 2.1: New documentation

But online, no matter what version you choose, even going all the way back to .NET Framework 4.5, you are presented with the new documentation.
![image](https://user-images.githubusercontent.com/22459191/151404691-a70f0ec3-fecb-4d56-9fb8-098c93a4d96c.png)

I'm not sure when the online documentation was updated, but given that the serilog-sinks-syslog project targets .NET Framework 4.6, .NET Standard 2.0, and .NET Core 3.1, the Intellisense won't show the new documentation unless you specifically have the .NET Core target selected. So that's is probably the whole reason why we're in this situation in the first place. (For a project that references serilog-sinks-syslog, the Intellisense documentation will be dependent upon that projects target framework, which I imagine you are using .NET Core/5/6.)

Next question is, where did you get the serilog-sinks-syslog example that you show a screen shot of? While it looks almost identical to part of the sample program:
https://github.com/IonxSolutions/serilog-sinks-syslog/blob/v2.0.6/sample/Serilog.Sinks.Syslog.Sample/Program.cs#L44

I have no idea where the documentation you have in the screen shot is coming from. For example, if I search the entire serilog-sinks-syslog solution for "defines which protocols can be negotiated", I get zero results.

Given that, I'm not sure it's worth doing anything until possibly a major release and then potentially making it a breaking change. @colin-anderson, unless you feel strongly about it and want to try introducing a change now.

**Comment by colin-anderson** [comment]
> The part about None is chopped off

Not sure how it ended up like that, but going through the history I can see it was like that from the get-go - it's fixed now 😃 

> VS didn't display any intellisense for SecureProtocols

At least for `SyslogTcpConfig`, it is [there](https://github.com/IonxSolutions/serilog-sinks-syslog/blob/master/src/Serilog.Sinks.Syslog/Sinks/Settings/SyslogTcpConfig.cs#L49).

**Comment by colin-anderson** [comment]
@Kevin-Andrew 

> Given that, I'm not sure it's worth doing anything until possibly a major release and then potentially making it a breaking change

Aye, agreed, I'll leave this issue open to see if anyone else has a problem with `SslProtocols`, and we'll consider changing the behaviour at a major release.

**Comment by CaringDev** [comment]
+1 for "anyone else has a problem"... given (CA5398)[https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca5398] I too would like to see this change. Also, I'd not go for ORing protocols as this would prevent eventual, automatic upgrades but would require code changes once new protocols are introduced.

**Comment by Edeendt** [comment]
I am also interested in the Ssl.Auto configuration, because i am afraid that hard coded ssl version will not pass a security audit.

**Comment by antibarbie** [comment]
Just to add 2cents, we also need to pass security audits, and are interested by the SllProtocols.None options. 
Ssl.Auto... yeah but others people are using None for automatic, so it's not great for lisibility.

Maybe an extension method **.WithUseTls()** that force UseTls would fix this without inducing compatibility issues.

**Comment by Kevin-Andrew** [comment]
Given [CA5398](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca5398) and the discussion thus far, if people are wanting a change, then I think at this point we look into making a breaking change, as that will hopefully be the cleanest in the long run.

@colin-anderson , what are your thought about doing that, and what would you want to do with the versioning?

**Comment by colin-anderson** [comment]
@Kevin-Andrew aye, it sounds like it's time to get this fixed 😅 As for versioning, I think we'd go to v3.0.0 after this.

As far an details go, are you happy with simply adding a `bool TlsEnabled` flag on the `SyslogTcpConfig` class (and making appropriate changes to the extensions, obv)?

**Comment by Kevin-Andrew** [comment]
@colin-anderson 
> ... are you happy with simply adding a `bool TlsEnabled` flag on the `SyslogTcpConfig` class (and making appropriate changes to the extensions, obv)?

Yeah, that is what I was thinking. The current `secureProtocols` parameter would be replaced by this new flag. Under the hood, if the new flag is enabled, we'd simply use the .NET `SslProtocols.None` option to let the OS negotiate the most secure option. If people want control over what SSL/TLS versions to support, then they will have to do so at the OS level, because ultimately the OS has the final say.

I'm not sure if there is a need to expose another parameter to allow our users to be able to explicitly set the .NET `SslProtocols` option. If there ever is a need, we could add it, but for starters, I think we try to just keep it simple.

**Comment by colin-anderson** [comment]
@Kevin-Andrew I was assuming we'd leave the `SecureProtocols` in place, defaulted to `SslProtocols.None`, in addition to a new `TlsEnabled`. But on reflection, I'm struggling to come up with a use case!

@Edeendt @CaringDev @antibarbie you folks happy if the `SecureProtocols` config option goes away, and we always leave it up to the OS to negotiate a protocol with the server, or do you have some need for it?

**Comment by CaringDev** [comment]
> happy if the `SecureProtocols` config option goes away, and we always leave it up to the OS to negotiate a protocol with the server

for now this is 👌and I totally agree with

> try to just keep it simple

👍

**Comment by antibarbie** [comment]
> I was assuming we'd leave the `SecureProtocols` in place, defaulted to `SslProtocols.None`, 
> in addition to a new `TlsEnabled`. But on reflection, I'm struggling to come up with a use case!

Yes, the new `TlsEnabled `is the only really _needed_ thing. 

> you folks happy if the `SecureProtocols` config option goes away, 
> and we always leave it up to the OS to negotiate a protocol with the server, or do you have some need for it?

Indeed, we have no use for it.

But if you wanted to keep compatibility (I think I read something about compatibility above in the comments) maybe you can keep it anyway.

Thanks for your reactivity.

**Comment by Edeendt** [comment]
> you folks happy if the `SecureProtocols` config option goes away, and we always leave it up to the OS to negotiate a protocol with the server, or do you have some need for it?

I am happy with any fix that allows the OS to negotiate the protocol.
However, you don't need to do breaking change for me, even though I can also understand that you want to keep it simple.

**Comment by colin-anderson** [comment]
Resolved by #63, new release coming soon

**Comment by colin-anderson** [comment]
New release is now available on NuGet: https://www.nuget.org/packages/Serilog.Sinks.SyslogMessages/3.0.0

**Comment by antibarbie** [comment]
Just for reference, we found a "corner case" where a C++/Cli program targetting .NET 4.7.2 or 4.8 would need the following registry entries, in order for option "None" to function properly. I could not find a clear reason for that, since Microsoft documentation says "only NET 3.5 needs that".

`````
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\.NETFramework\v4.0.30319]
"SchUseStrongCrypto"=dword:00000001
"SystemDefaultTlsVersions"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\.NETFramework\v4.0.30319]
"SchUseStrongCrypto"=dword:00000001
"SystemDefaultTlsVersions"=dword:00000001
``````

Just posting this in case it would help somebody else, hit by the same bug.

**Comment by colin-anderson** [comment]
@antibarbie thanks for sharing - it's the kind of obscure information that's guaranteed to save someone a lot of time at some point!
